### PR TITLE
Enhance last query, support non single base series

### PR DESCRIPTION
**Pull request by Beyyes**
## Description

see design doc in: https://timechor.feishu.cn/docx/KoqTdQKTlo1JcOx41IQcwajKn2g

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java [hunk 1]
```diff
@@ -495,20 +495,28 @@ private Analysis analyzeLastQuery(
 
   private void analyzeLastSource(
       Analysis analysis, List<Expression> selectExpressions, ISchemaTree schemaTree) {
-    Set<Expression> sourceExpressions;
-
-    sourceExpressions = new LinkedHashSet<>();
+    Set<Expression> sourceExpressions = new LinkedHashSet<>();
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java:498]: If we still need this?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -301,6 +313,54 @@ public LogicalPlanBuilder planLast(
       }
     }
 
+    if (lastQueryNonWriteViewExpressions != null) {
+      for (Expression expression : lastQueryNonWriteViewExpressions) {
+        Set<Expression> sourceTransformExpressions =
+            new HashSet<>(Collections.singletonList(expression));
+        LogicalPlanBuilder planBuilder = new LogicalPlanBuilder(analysis, context);
+        Set<Expression> sources = new LinkedHashSet<>(searchSourceExpressions(expression));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java:321]: we do searchSourceExpressions twice here, first time is in `analyzeLastSource`

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -301,6 +313,54 @@ public LogicalPlanBuilder planLast(
...
+
+        planBuilder =
+            planBuilder
+                .planRawDataSource(
+                    sources,
+                    queryStatement.getResultTimeOrder(),
+                    analysis.getGlobalTimeFilter(),
+                    analysis.isLastLevelUseWildcard())
+                .planWhereAndSourceTransform(
+                    null,
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java:340]: does timefilter push down?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java [hunk 4]
```diff
@@ -301,6 +313,54 @@ public LogicalPlanBuilder planLast(
       }
     }
 
+    if (lastQueryNonWriteViewExpressions != null) {
+      for (Expression expression : lastQueryNonWriteViewExpressions) {
+        Set<Expression> sourceTransformExpressions =
+            new HashSet<>(Collections.singletonList(expression));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java:319]:
```suggestion
            Collections.singleton(expression);
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java [hunk 5]
```diff
@@ -301,6 +313,54 @@ public LogicalPlanBuilder planLast(
...
+                    analysis.getGroupByTimeParameter(),
+                    analysis.getGroupByParameter(),
+                    false,
+                    AggregationStep.SINGLE,
+                    Ordering.ASC);
+
+        LastQueryTransformNode transformNode =
+            new LastQueryTransformNode(
+                context.getQueryId().genPlanNodeId(),
+                Arrays.asList(planBuilder.getRoot()),
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java:357]:
```suggestion
                Collections.singletonList(planBuilder.getRoot()),
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/OperatorTreeGenerator.java [hunk 6]
```diff
@@ -2340,6 +2342,10 @@ public Operator visitAlignedLastQueryScan(
 
   @Override
   public Operator visitLastQuery(LastQueryNode node, LocalExecutionPlanContext context) {
+    if (node.getChildren().size() == 1
```
[the remark this answers is not in the document]
Comment (JackieTien97) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/OperatorTreeGenerator.java:2345]: What's this for?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/OperatorTreeGenerator.java [hunk 7]
```diff
@@ -2451,6 +2457,23 @@ public Operator visitLastQueryCollect(
     return new LastQueryCollectOperator(operatorContext, children);
   }
 
+  @Override
+  public Operator visitLastQueryTransform(
+      LastQueryTransformNode node, LocalExecutionPlanContext context) {
+    List<Operator> children = dealWithConsumeChildrenOneByOneNode(node, context);
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/OperatorTreeGenerator.java:2463]: No need to call this if there is only one child. Just call like the following:

`node.getChild().accept(this, context);`

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/SubPlanTypeExtractor.java [hunk 8]
```diff
@@ -128,17 +129,22 @@ public Void visitAlignedLastQueryScan(AlignedLastQueryScanNode node, Void contex
 
     @Override
     public Void visitLastQuery(LastQueryNode node, Void context) {
-      return null;
+      return visitPlan(node, context);
```
[the remark this answers is not in the document]
Comment (JackieTien97) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/SubPlanTypeExtractor.java:132]: That's right, we should never influence the performance of normal last query.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java (hunk 5 above)
Comment (Beyyes) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java:357]: Using Collections.singletonList may meets error when calling "clone" method of PlanNode, I will retry again.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java (hunk 3 above)
Comment (Beyyes) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanBuilder.java:340]: I'm not sure, is it not enough to put it in planRawDataSource?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanVisitor.java [hunk 9]
```diff
@@ -620,7 +624,7 @@ public PlanNode visitShowTimeSeries(
         && 0 != analysis.getDataPartitionInfo().getDataPartitionMap().size()) {
       PlanNode lastPlanNode =
           new LogicalPlanBuilder(analysis, context)
-              .planLast(analysis.getSourceExpressions(), analysis.getGlobalTimeFilter(), null)
+              .planLast(analysis, analysis.getGlobalTimeFilter(), null, ASC, ZoneId.systemDefault())
```
Comment (Beyyes) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/LogicalPlanVisitor.java:627]: ASC and systemDefaultZoneId in show latest timeseries

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**